### ISSUE TITLE
mongo-connector stops working when active mongos and node go offline

### ISSUE THREAD

**Issue by vimuston** [mongodb-labs/mongo-connector#549]
Our environment has following MongoDB cluster setup:
**Site 1:**
config1: config & mongos
config2: config & mongos
node1: rs1 & rs2
node2: rs1 & rs2

**Site 3:**
config3: config & mongos
node3: rs1 & rs2

mongo-connector **mainAddress** is "config1:27017,config2:27017,config3:27017"

On the weekend **Site 1** went down and node3 on **Site 2** was elected primary. During this time mongo-connector raised an AutoReconnect and stopped indexing documents. The mongo-connector process is still alive, but nothing is getting inserted to Solr.

Judging from where the exception was raised from, it would seem that the OplogThread noticed the disconnection and the control returned to Connector. However, the mongos Connector was connected to (this I am guessing but it would make sense) is also down, because it resides on the same site. In this scenario, the main connection should probably be reconnected?

[mongo-connector.txt](https://github.com/mongodb-labs/mongo-connector/files/521814/mongo-connector.txt)

Edit: Including the traceback:
```python
2016-10-09 00:53:06,512 [ERROR] mongo_connector.util:92 - Fatal Exception
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/mongo_connector-2.4.1-py2.6.egg/mongo_connector/util.py", line 90, in wrapped
    func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/mongo_connector-2.4.1-py2.6.egg/mongo_connector/connector.py", line 316, in run
    for shard_doc in main_conn['config']['shards'].find():
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1149, in next
    if len(self.__data) or self._refresh():
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1081, in _refresh
    self.__codec_options.uuid_representation))
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 996, in __send_message
    res = client._send_message_with_response(message, **kwargs)
  File "build/bdist.linux-x86_64/egg/pymongo/mongo_client.py", line 1366, in _send_message_with_response
    raise AutoReconnect(str(e))
AutoReconnect: [Errno 113] No route to host
```

**Comment by ShaneHarvey** [comment]
Can I ask how you are using MongoConnector? The mainAddress for MongoConnector should not be a config server. MongoConnector requires passing mongos addresses with a sharded cluster, eg:
`$ mongo-connector -m mongoshost1:1234,mongoshost2:1234`

**Comment by vimuston** [comment]
Sorry that was my mistake -- the config nodes also host the mongos instances and that is what mongo-connector is connected to.

**Comment by vimuston** [comment]
Upon retesting the scenario, it seems that mongo-connector does in fact quit when the mongos connection fails. I am no longer sure whether this was the scenario in the original report as well, and the 'process is still running' was a mistake in the report. I cannot reproduce a scenario where the mongos connection fails but the oplog threads won't join -- this would cause the described behaviour, no?

**Comment by vimuston** [comment]
@ShaneHarvey Are you currently working on this? I noticed #561 seems to fix a similar kind of issue. I have a fix proposal for the connector.py part: wrap the 'config.shards' loop in try/except and catch pymongo.errors.AutoReconnect, so that the next connection attempt reconnects correctly. I tested it in our environment by shutting down the active mongos and one of the oplog mongod instances. mongo-connector seemed to recover correctly.
